src/main/java/heronarts/lx/midi/surface/APC40Mk2.java

```diff
@@ -291,50 +317,88 @@ void registerNext() {
-        if (this.device instanceof LXEffect) {
-          this.effect = (LXEffect) this.device;
-          this.effect.enabled.addListener(this);
-        } else if (this.device instanceof LXPattern) {
-          this.pattern = (LXPattern) this.device;
+      if (this.device == device) {
+        return;
+      }
+      unregister();
+      this.device = device;
+
+      if (this.device == null) {
+        clearKnobsAfter(0);
+        return;
+      } else if (this.device instanceof LXEffect) {
+        this.effect = (LXEffect) this.device;
+        this.effect.enabled.addListener(this);
+      } else if (this.device instanceof LXPattern) {
+        this.pattern = (LXPattern) this.device;
+      }
+
+      this.device.controlSurfaceSemaphore.increment();
+
+      boolean isEnabled = false;
+      if (this.effect != null) {
+        isEnabled = this.effect.isEnabled();
+      } else if (this.pattern != null) {
+        isEnabled = this.pattern == ((LXChannel) this.channel).getActivePattern();
+      }
+      sendNoteOn(0, DEVICE_ON_OFF, isEnabled ? LED_ON : LED_OFF);
+
+      int i = 0;
+      LXListenableParameter currentParentParameter = null;
+      for (LXListenableNormalizedParameter parameter : this.device.getRemoteControls()) {
+        if (parameter == null) {
+          this.knobs[i] = null;
+          sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
+          continue;
         }
 
-        int i = 0;
-        boolean isEnabled = false;
-        if (this.device != null) {
-          if (this.effect != null) {
-            isEnabled = this.effect.isEnabled();
-          } else if (this.pattern != null) {
-            isEnabled = this.pattern == ((LXChannel) this.channel).getActivePattern();
-          }
-          for (LXListenableNormalizedParameter parameter : this.device.getRemoteControls()) {
-            if (i >= this.knobs.length) {
-              break;
-            }
-            this.knobs[i] = parameter;
-            if (parameter != null) {
-              parameter.addListener(this);
-              sendControlChange(0, DEVICE_KNOB_STYLE + i, parameter.getPolarity() == LXParameter.Polarity.BIPOLAR ? LED_STYLE_BIPOLAR : LED_STYLE_UNIPOLAR);
-              double normalized = (parameter instanceof CompoundParameter) ?
-                ((CompoundParameter) parameter).getBaseNormalized() :
-                parameter.getNormalized();
-              sendControlChange(0, DEVICE_KNOB + i, (int) (normalized * 127));
-            } else {
-              sendControlChange(0, DEVICE_KNOB_STYLE + i, LED_STYLE_OFF);
-            }
-            ++i;
+        if (i >= this.knobs.length) {
+          // Subtle: Even if all the knobs are full, we need to keep adding
+          // listeners for all children of the final knob. So only break
+          // if we're up to a parameter with no parent or a different parent.
+          if (parameter.parentParameter == null || parameter.parentParameter != currentParentParameter) {
```

Comment:
`AggregateParameter parentParameter = parameter.getParentParameter();`

Grab that once via accessor and use it for all the references from here down.

Reply:
Done